dali/operators/io/file/file_read.cc

```diff
@@ -0,0 +1,110 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/pipeline/operator/checkpointing/stateless_operator.h"
+#include "dali/core/util.h"
+#include "dali/util/file.h"
+
+namespace dali {
+namespace io {
+
+class FileRead : public StatelessOperator<CPUBackend> {
+ public:
+  explicit FileRead(const OpSpec &spec)
+      : StatelessOperator<CPUBackend>(spec),
+        dont_use_mmap_(spec.GetArgument<bool>("dont_use_mmap")),
+        use_o_direct_(spec.GetArgument<bool>("use_o_direct")) {}
+
+  bool CanInferOutputs() const override { return true; }
+
+  bool SetupImpl(std::vector<OutputDesc> &output_descs, const Workspace &ws) override {
+    const auto &filepaths = ws.Input<CPUBackend>(0);
+    DALI_ENFORCE(filepaths.type() == DALI_UINT8 || filepaths.type() == DALI_INT8,
+                 "Unexpected input data type");
+    auto nsamples = filepaths.shape().num_samples();
+    output_descs.resize(1);
+    output_descs[0].type = DALI_UINT8;
+    output_descs[0].shape.resize(nsamples, 1);
+    filenames_.resize(nsamples);
+    files_.resize(nsamples);
+    for (int i = 0; i < nsamples; i++) {
+      size_t filename_len = filepaths.tensor_shape_span(i)[0];
+      filenames_[i].resize(filename_len + 1);
+      std::memcpy(&filenames_[i][0], filepaths.raw_tensor(i), filename_len);
+      filenames_[i][filename_len] = '\0';
+
+      FileStream::Options opts;
+      opts.use_mmap = !dont_use_mmap_;
+      opts.use_odirect = use_o_direct_;
+      opts.read_ahead = false;
+      files_[i] = FileStream::Open(filenames_[i], opts);
+      output_descs[0].shape.tensor_shape_span(i)[0] = files_[i]->Size();
+    }
+    return true;
+  }
+
+  void RunImpl(Workspace &ws) override {
+    auto &filenames = ws.Input<CPUBackend>(0);
+    auto &file_contents = ws.Output<CPUBackend>(0);
+    auto &tp = ws.GetThreadPool();
+    int nthreads = tp.NumThreads();
+    for (size_t idx = 0; idx < filenames_.size(); idx++) {
+      size_t file_size = file_contents.shape().tensor_size(idx);
+      tp.AddWork(
+          [&, file_size, idx](int tid) {
+            auto &stream = files_[idx];
+            auto read_nbytes =
+                stream->Read(file_contents.mutable_tensor<uint8_t>(idx), file_size);
+            DALI_ENFORCE(read_nbytes == file_size,
+                         make_string("Expected to read ", file_size, " bytes, but got only ",
+                                     read_nbytes, " bytes."));
+            stream->Close();
+            stream.reset();
+          },
+          file_size);
+    }
+    tp.RunAll();
+  }
+
+ private:
+  bool dont_use_mmap_;
+  bool use_o_direct_;
+  std::vector<std::string> filenames_;
+  std::vector<std::unique_ptr<FileStream>> files_;
+};
+
+}  // namespace io
+
+DALI_REGISTER_OPERATOR(io__file__Read, io::FileRead, CPU);
+DALI_SCHEMA(io__file__Read)
```

Comment:
I just want to keep this in the history of conversation, but I would consider `io.read.file` so the potential "readers" for other containers would all land in one module.